cdap-app-fabric/src/main/java/io/cdap/cdap/reporting/ProgramHeartbeatTable.java

```diff
@@ -99,49 +103,95 @@ public void deleteAll() throws IOException {
+      List<Field<?>> startRowKey = getScanKey(namespace, rangeStart);
+      List<Field<?>> endRowKey = getScanKey(namespace, runningOnTimestamp);
+      Range range = Range.create(startRowKey, Range.Bound.INCLUSIVE, endRowKey, Range.Bound.EXCLUSIVE);
+      ranges.add(range);
     }
+    performMultiScanAddToList(ranges, resultRunRecordList, (rrd) -> rrd.getStatus() == ProgramRunStatus.RUNNING);
     return resultRunRecordList;
   }
 
   /**
-   * Scan is executed based on the given startRowKey and endRowKey, for each of the scanned rows, we maintain
-   * the latest {@link RunRecordDetail} identified by its {@link ProgramRunId} in a map. Finally after scan is
-   * complete add the runrecords to the result list
+   * Scan is executed based on the given startRowKey and endRowKey, for each of the scanned rows, we maintain the latest
+   * {@link RunRecordDetail} with a status of RUNNING, identified by its {@link ProgramRunId} in a map. If a record has
+   * a status other than RUNNING, is removed from the result set. Finally after scan is complete add the runrecords to
+   * the result list
+   *
+   * @param ranges         the ranges to query
+   * @param runRecordMetas result list to which the run records to be added
+   */
+  private void performMultiScanAddToList(Collection<Range> ranges,
+                                         List<RunRecordDetail> runRecordMetas) throws IOException {
+    performMultiScanAddToList(ranges, runRecordMetas, null);
+  }
+
+  /**
+   * Scan is executed based on the given startRowKey and endRowKey, for each of the scanned rows, we maintain the latest
+   * {@link RunRecordDetail} with a status of RUNNING, identified by its {@link ProgramRunId} in a map. If a record has
+   * a status other than RUNNING, is removed from the result set. Finally after scan is complete add the runrecords to
+   * the result list
    *
-   * @param startRowKey byte array used as start row key in scan
-   * @param endRowKey byte array used as end row key in scan
+   * @param ranges         the ranges to query
    * @param runRecordMetas result list to which the run records to be added
+   * @param predicate      the predicate used to filter results in the output.
    */
-  private void performScanAddToList(List<Field<?>> startRowKey, List<Field<?>> endRowKey,
-                                    List<RunRecordDetail> runRecordMetas) throws IOException {
-    Map<ProgramRunId, RunRecordDetail> latestRunRecords = new LinkedHashMap<>();;
-    try (CloseableIterator<StructuredRow> iterator =
-      table.scan(Range.create(startRowKey, Range.Bound.INCLUSIVE, endRowKey, Range.Bound.EXCLUSIVE),
-                 Integer.MAX_VALUE)) {
+  private void performMultiScanAddToList(Collection<Range> ranges,
+                                         List<RunRecordDetail> runRecordMetas,
+                                         @Nullable Function<RunRecordDetail, Boolean> predicate) throws IOException {
```

Comment:
a little cleaner not to allow a null and have the other method pass in `x -> true`

Reply:
Clever. I'll make this change.

Reply:
I have implemented this change.